### PR TITLE
trend model

### PR DESCRIPTION
**Pull request by krsnapaudel**
Implemented a trend model class:
* The model constructor expects names of columns containing x values and y values.
* The model expects years in x_cols and yields in y_cols. Since these features are created outside the class, there is no work done during training.
* The linear trend computation is kept separate to allow overriding this.

### CODE REVIEW

**Comment by krsnapaudel** [comment]
Some additional changes not related to trend model, but to formatting with `black`.

**Comment by ritviksahajpal** [comment]
@krsnapaudel seems like you have the trend part covered, perhaps including a quadratic model apart from the linear model could be an option, but it is not needed.

**Comment by krsnapaudel** [comment]
@ritviksahajpal : Yes. If you have a formula for quadratic trend, please suggest.

**Comment by ritviksahajpal** [comment]
Something like this:


""" column_name: string name of the column to detrend. """
from statsmodels.regression.linear_model import OLS
from statsmodels.tools.tools import add_constant

""" Add code to exclude years in the future """
df["t"] = np.array(df["Harvest Year"])

""" Quadratic trend model """
X_quad = add_constant(np.column_stack((df["t"], df["t"] ** 2)))
quad_model = OLS(df[column_name], X_quad).fit()

detrended = df[column_name] - quad_model.predict(X_quad)


I agree that the code will need to ensure that only years prior to forecast year are used

**Comment by krsnapaudel** [comment]
@ritviksahajpal, @michielkallenberg : Not sure what the best way to alternate among trend estimation methods. I made it a constructor argument. Please share your thoughts.

File: models/trend_model.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        """
+
+        predictions = np.zeros((len(X), 1))
+        for i, item in enumerate(X):
+            trend_x = [item[y] for y in self._x_cols]
+            trend_y = [item[c] for c in self._y_cols]
+
+            predictions[i] = self._trend_fn(trend_x, trend_y, item[KEY_YEAR])
+
+        print(predictions)
```
Comment (michielkallenberg) [inline models/trend_model.py:52]: Remove debugging

**Review by michielkallenberg — COMMENTED** [review]
[disclaimer: I may be very much wrong here] Didn't inspect all details, but I have a general question: what should we do if we have a training set and a test set? I figure that the proposed code is not meant for training with a separate training set. It just takes a test set it seems. So when predicting a year from the test set, how do we go about that. Do we use "surrounding years" from the *test set* as features, or do we take those years from the training set?
Can't we store learnt parameters as I figure was done in a previous implementation?

**Comment by krsnapaudel** [comment]
@michielkallenberg : The previous implementation also did not learn anything during fit. See
https://github.com/BigDataWUR/AgML-crop-yield-forecasting/blob/krsnapaudel-python-package/models/trend_models.py
It just copied the data.

The idea is predict will get trend features. In this implementation, they come as years and yields for those years. The caller is expected to design the features correctly.

**Comment by krsnapaudel** [comment]
The most recent code implements one trend for the whole country.
There is a panel data with fixed effects approach, but assumes parallel trend for all locations.
https://matheusfacure.github.io/python-causality-handbook/14-Panel-Data-and-Fixed-Effects.html

Should we train a separate model for each location? @michielkallenberg, @ritviksahajpal : Suggest what's appropriate.

**Comment by ritviksahajpal** [comment]
You will likely need a separate model for each region unless you implement a mixed-effects model.

**Comment by michielkallenberg** [comment]
> The most recent code implements one trend for the whole country. There is a panel data with fixed effects approach, but assumes parallel trend for all locations. https://matheusfacure.github.io/python-causality-handbook/14-Panel-Data-and-Fixed-Effects.html
> 
> Should we train a separate model for each location? @michielkallenberg, @ritviksahajpal : Suggest what's appropriate.

An easy "hack" would be to use one-hot encoding for location, I figure

**Comment by ritviksahajpal** [comment]
@michielkallenberg , models like random forest in scikit-learn do not deal well with high cardinality features (https://stackoverflow.com/questions/24715230/can-sklearn-random-forest-directly-handle-categorical-features), I realize not all models have this limitation, but something to take into consideration

**Comment by michielkallenberg** [comment]
> @michielkallenberg , models like random forest in scikit-learn do not deal well with high cardinality features (https://stackoverflow.com/questions/24715230/can-sklearn-random-forest-directly-handle-categorical-features), I realize not all models have this limitation, but something to take into consideration, (U.S. data has ~3,000 counties)

OK thanks for your insights. I figure one-hot encoding probably is a bad idea anyway since interaction effects (i.e. a truly different model for each location) are typically omitted in these regression models I figure.

**Comment by krsnapaudel** [comment]
I am also leaning towards a per-admin-region model.

**Comment by krsnapaudel** [comment]
I have made the average model work the same way as trend model now. Removed lowess since prediction with lowess seems complicated.

**Review by michielkallenberg — APPROVED** [review]
LGTM.
